### PR TITLE
Update Stacked Autoencoder and Denoising functionality

### PR DESCRIPTION
**Pull request by collinleiber**
- Update the implementation of the Stacked Autoencoder to match the structure of the ClustPy package
- Add possibility to specify a corruption function in all neural networks. This is required to perform a denoising training

See #24

### CODE REVIEW

File: clustpy/deep/neural_networks/_abstract_autoencoder.py [hunk 1]
```diff
@@ -81,6 +84,18 @@ def forward(self, x: torch.Tensor) -> torch.Tensor:
         forwarded = self.block(x)
         return forwarded
 
+    def get_linear_layers(self) -> np.ndarray:
```
Comment (Lumik7) [inline clustpy/deep/neural_networks/_abstract_autoencoder.py:87]: instead of inferring this, you could add ids during the construction of the layers in FullyConnectedBlock?

File: clustpy/deep/neural_networks/variational_autoencoder.py [hunk 2]
```diff
@@ -162,7 +169,9 @@ def loss(self, batch: list, ssl_loss_fn: torch.nn.modules.loss._Loss, device: to
         """
         assert type(batch) is list, "batch must come from a dataloader and therefore be of type list"
         batch_data = batch[1].to(device)
-        z, q_mean, q_logvar, reconstruction = self.forward(batch_data)
+        batch_data_adj = batch_data if corruption_fn is None else corruption_fn(batch_data)
```
Comment (Lumik7) [inline clustpy/deep/neural_networks/variational_autoencoder.py:172]: same problem as described in comment above

File: clustpy/deep/neural_networks/stacked_autoencoder.py [hunk 3]
```diff
@@ -1,258 +1,232 @@
...
+            for _ in range(n_epochs_per_layer):
+                total_loss = 0
+                for batch in dataloader:
+                    input_data = batch[1].to(device)
+                    with torch.no_grad():
+                        # encode batch using already trained layers including non-linearity functions etc
+                        for encode_layer in range(encoder_layer_to_train):
+                            input_data = self.encoder.block[encode_layer](input_data)
+                    # Calculate loss regarding current layer
+                    input_data_adj = input_data if corruption_fn is None else corruption_fn(input_data)
```
Comment (Lumik7) [inline clustpy/deep/neural_networks/stacked_autoencoder.py:138]: Same problem as described in comment above

File: clustpy/deep/neural_networks/stacked_autoencoder.py [hunk 4]
```diff
@@ -1,258 +1,232 @@
...
+                    encoded = self.encoder.block[encoder_layer_to_train](input_data_adj)
```
Comment (Lumik7) [inline clustpy/deep/neural_networks/stacked_autoencoder.py:139]: how would batchnorm be handled in this setting?

**Review by Lumik7 — CHANGES_REQUESTED** [review]
Looks good, but the corruption_fn design choices need to be thought through again.

**Review by Lumik7 — APPROVED** [review]
